### PR TITLE
Use error message returned by isImplicitlyConvertible

### PR DESCRIPTION
**Pull request by Marenz**
Early draft for https://github.com/ethereum/solidity/issues/4128
I haven't changed all calls of `isImplicitlyConvertible` yet and I still have to add  the actual error messages.

Early feedback is welcome.

### CODE REVIEW

**Comment by Marenz** [comment]
Okay, this is ready for review.
The function doesn't return any errors yet.

File: liblangutil/ErrorReporter.h [hunk 1]
```diff
@@ -87,6 +87,11 @@ class ErrorReporter
 
 	void typeError(SourceLocation const& _location, std::string const& _description);
 
+	void typeError(
+		SourceLocation const& _location,
+		std::initializer_list<std::string const> const& _descriptions
```
Comment (leonardoalt) [inline liblangutil/ErrorReporter.h:92]: Why is this necessary?

Comment (Marenz) [reply liblangutil/ErrorReporter.h:92]: Otherwise instead of doing

```
		m_errorReporter.typeError(
			arguments.front()->location(),
			{"Invalid type for argument in function call. "
			"Invalid implicit conversion from " +
			type(*arguments.front())->toString() +
			" to bytes memory requested.",
			details.message()}
		);
```
I would need to do

```
		m_errorReporter.typeError(
			arguments.front()->location(),
			"Invalid type for argument in function call. "
			"Invalid implicit conversion from " +
			type(*arguments.front())->toString() +
			" to bytes memory requested." +
			(details.message().empty() ? "" : " " + details.message())
		);
```

for every single occurrence which imo looks rather messy and not easy to read.
This approach seemed the cleanest way to me.

Comment (chriseth) [reply liblangutil/ErrorReporter.h:92]: If we decide to do it like that, I would actually prefer a template function with variable arguments instead of an initializer list - the `{` look a bit weird in the argument list.

Comment (Marenz) [reply liblangutil/ErrorReporter.h:92]: I opted for the initializer list because it provides type safety and.. well templates always cost compile time, too.

File: libsolidity/analysis/TypeChecker.cpp [hunk 2]
```diff
@@ -139,13 +139,19 @@ TypePointers TypeChecker::typeCheckABIDecodeAndRetrieveReturnType(FunctionCall c
 			toString(arguments.size()) +
 			" were provided."
 		);
-	if (arguments.size() >= 1 && !type(*arguments.front())->isImplicitlyConvertibleTo(ArrayType::bytesMemory()))
+
+	BoolResult details(false);
```
Comment (erak) [inline libsolidity/analysis/TypeChecker.cpp:143]:
```suggestion
	BoolResult details{false};
```

File: liblangutil/ErrorReporter.h (hunk 1 above)
Comment (erak) [reply liblangutil/ErrorReporter.h:92]: I'd also vote for implementing it with a variadic template. The initializer list does indeed look a bit off.

File: liblangutil/ErrorReporter.cpp [hunk 3]
```diff
@@ -198,6 +198,30 @@ void ErrorReporter::typeError(SourceLocation const& _location, SecondarySourceLo
 	);
 }
 
+void ErrorReporter::typeError(SourceLocation const& _location, std::initializer_list<std::string const> const& _descriptions)
+{
+	std::string errorStr;
```
Comment (erak) [inline liblangutil/ErrorReporter.cpp:203]: Could we use `std::transform` here?

```
std::vector<std::string> allErrors;
std::transform(s.begin(), s.end(), std::back_inserter(allErrros), 
    [](string error) -> std::string { return error; });
```

Comment (Marenz) [reply liblangutil/ErrorReporter.cpp:203]: how would that help?

Comment (erak) [reply liblangutil/ErrorReporter.cpp:203]: Aeh, you do not have to fiddle with iterators in a raw loop?

Comment (Marenz) [reply liblangutil/ErrorReporter.cpp:203]: but you're simply concatenating the strings in your code.. We need logic based on whether it's the last element or not to know if we should append a whitespace or not between the sentences..

Comment (leonardoalt) [reply liblangutil/ErrorReporter.cpp:203]: That could go in the lambda. I'm not sure which one would be more readable though.

**Comment by Marenz** [comment]
I am gonna close this for now as other issues take precedence. Will take it up again when the time is right.

File: liblangutil/ErrorReporter.h (hunk 1 above)
Comment (chriseth) [reply liblangutil/ErrorReporter.h:92]: @Marenz if this clashes with something, you can always add another function with a different name.

Comment (Marenz) [reply liblangutil/ErrorReporter.h:92]: Well, that would be an easy way to work around my current problem.
